torch/_functorch/_aot_autograd/autograd_cache.py

```diff
@@ -63,11 +63,36 @@ def check_node_safe(node: Node):
     """
     Checks that the node only uses supported operators. We are starting with very
     conservative cacheability constraints, and incrementally adding more support as we expand.
+
+    [Note: AOTAutograd Cacheability checks]
+    - Our cache key is computed from the FX graph produced by Dynamo and the input example values
+    - A node is "safe" if the same cache key results in a compiled artifact that has the same behavior
+        (i.e, the set of inputs that go into our cache key is sufficient to distinguish its behavior)
+
+    To accomplish this safety check, we consider the following functions to be safe:
+        - Public functions under modules torch, torch.functional, and torch.nn.functional: these are
+        allowed in the graph by dynamo, so we can assume they are safe to cache.
+        - method calls on base tensor types
+        - Any call_module that dynamo deemed safe to allow AOTAutograd to trace
+        - Non callable nodes, such as placeholder, output, get_attr
+
+    The test suite test_aot_autograd_cache.py::AOTAutogradCachePicklerTests tries its best to fully cover/specify this behavior.
     """
+    SAFE_TORCH_MODULES = ("torch.functional", "torch.nn.functional", "torch")
+
+    def is_public_torch_api(target):
+        # Don't blindly allow private functions in the torch namespace
+        is_private = target.__name__.startswith("_")
+        return (
+            getattr(target, "__module__", None) in SAFE_TORCH_MODULES and not is_private
```

Comment:
The thing I am struggling with right now is that while it may be true that all APIs in these modules today are safe, nothing prevents someone from adding a new unsafe API and introducing silent incorrectness to the cache. For example, `torch._higher_order_ops.triton_kernel_wrapper_mutation` is an API that is not cache safe (it uses integers to refer to triton kernels), but we could easily make it a public API under torch.* some day.

Reply:
Agreed, I don't think we would want to blindly allow anything under `torch.*` ("we might arbitrarily move higher order ops in there" feels like a good example).

How do you feel about something like this to start? Allow things under `torch.*` to be cached if either:

(1) it is a `builtin_function_or_method` (ideally, dynamo would inline all torch -related things straight to their C binding but sadly not today)

(2) it is under `torch.nn.functional` or `torch.functional` (our assumption here is that everything under these modules are things that dynamo morally should be able to trace into C-bindings today, but it does not).

That would allow us to not blanket-allow everything else under `torch.*`, like higher order ops that we happen to make public

Reply:
Got it, I can do that; that said, I feel like it might kind of just punts the problem later, when there's a torch.* function that we actually want to support that doesn't trace into C. 

I wonder if there's a way we can check safety by being able to understand whether a function generates guards in dynamo. If there's an API we can use to check "are there no dynamo guards associated with this function", I think that would be a decent proxy for safety.

Reply:
I'll file an issue for a more general solution to this problem but for now, I will hardcode a few torch.* functions that are safe but don't bind to C (i.e. torch.einsum) in a list for this PR. 

I think a good solution here will involve a tag in native_functions.yaml for any torch functions that are implemented in python but are cache safe

Reply:
Update, there aren't any specific torch.* functions that actually need hard coding: torch.einsum is actually just an alias for troch.functional.einsum, so it's already covered (similar for others like it)